src/x/server/server_test.go

```diff
@@ -117,6 +130,142 @@ func TestServe(t *testing.T) {
 	s.Close()
 }
 
+func TestTLSPermissiveServerListenAndClose(t *testing.T) {
+	s, h, numAdded, numRemoved := testServer(testListenAddress, xtls.Permissive, false)
+
+	var (
+		numClients  = 9
+		expectedRes []string
+	)
+
+	err := s.ListenAndServe()
+	require.NoError(t, err)
+	listenAddr := s.listener.Addr().String()
+
+	for i := 0; i < numClients; i++ {
+		var conn net.Conn
+		var err error
+		if i%2 == 0 {
+			conn, err = net.Dial("tcp", listenAddr)
+		} else {
+			conn, err = tls.Dial("tcp", listenAddr, &tls.Config{InsecureSkipVerify: true})
+
+		}
+		require.NoError(t, err)
+
+		msg := fmt.Sprintf("msg%d", i)
+		expectedRes = append(expectedRes, msg)
+
+		_, err = conn.Write([]byte(msg))
+		require.NoError(t, err)
+	}
+
+	numChecks := 0
+	for h.called() < numClients && numChecks < 5 {
```

Comment:
Check out `assert.EventuallyWithT` for this; it's useful.

Reply:
To use that, we need to update Testify to the newest version, but honestly, I don't want to blow this pr more. We should probably update it in a separate pr.